3280 added nested belongs to rebased

rdallasgray opened the pull request:
This is a rebase of #3280 against the current master. We need this functionality and we're keen to see the PR merged. I understand the original author may not have had time to complete the rebase.

This commit also:
- Removes the extraneous `PostComment` model from the original PR and uses the existing `Tagging` model for specs
- Adds a feature spec for the breadcrumb links with nested `belongs_to` resources

smeyfroi commented:
+1  Yes!

timoschilling commented:
Hi, nice work. I am at the moment in the holidays. Reviewing thinks on mobile devises is hard. 

Can you please split this PR into a "fixed test" PR and after I merged that a fresh rebase. 
I will merge the "fixed tests" PR during 24h!

rdallasgray commented:
@timoschilling OK, thanks. New PR is here https://github.com/activeadmin/activeadmin/pull/4627

rdallasgray commented:
@timoschilling Rebased against master.

rdallasgray commented:
@timoschilling Rebased against master. I'm not sure how other people's tests are passing -- I had to pin `term-ansicolor` (brought in by `coveralls`) to get it past Ruby 1.9.

deivid-rodriguez commented:
@rdallasgray Yes, CI is broken again. Other people's tests are not passing, the green check just indicates that the build was green.. _when it was last run_, and that dependency was not broken at the time. Your fix is good.

There's another option though: switching test coverage service from coveralls to codeclimate, which has much less dependencies and it's already being used anyways for code quality. That way we could drop both `tins` & `term-ansicolor` dependencies. In case the maintainers are interested, I implemented that switch for RuboCop here: https://github.com/bbatsov/rubocop/commit/dfa1f9b6a2ad2f3d0d7e97a481665f68d4097f17. The reason was the same: broken `tins` dependency in old rubies.

deivid-rodriguez commented:
@rdallasgray Two comments:

* You now need to remove previous workarounds to get the tests green since the dependencies have been removed.
* The coverage report leads to a question. Is the `nested?` method now dead code, or is it just uncovered by tests?

varyonic commented on Gemfile:
```diff
@@ -64,10 +64,12 @@ end
+  gem 'coveralls', require: false                   # Test coverage website. Go to https://coveralls.io
+  gem 'tins', '~> 1.6.0', require: false            # Required by coveralls, > 1.6.0 removes support for Ruby 1.9
+  gem 'term-ansicolor', '~> 1.3.0', require: false  # Required by coveralls, > 1.4.0 removes support for Ruby 1.9
   gem 'json', '~> 1.8', require: false      # Required by simplecov, > 2 removes support for Ruby 1.9
```
I don't think this PR needs Gemfile changes?

varyonic commented on lib/active_admin/resource/routes.rb:
```diff
@@ -85,25 +85,29 @@ def route_name(resource_path_name, options = {})
         def nested?
-          resource.belongs_to? && resource.belongs_to_config.required?
+          !required_belongs_to.empty?
         end
```
As @deivid-rodriguez mentioned this looks like a dead method now.

rdallasgray commented:
OK, per comments from @varyonic and @Fivell:
- I've reverted the Gemfile to master 
- It does look like `nested?` was dead (the only mention of it in the codebase was the definition). I've removed it.

deivid-rodriguez commented:
I see duplicate commit names and https://github.com/activeadmin/activeadmin/pull/4618/commits/552fd5ea3c9bfb04ce725c4a8c1cd63de6fc9add is added and then reverted. I think cleaning up the commit history a bit could help the review.

rdallasgray commented:
@deivid-rodriguez This was taken over from #3280, so I've left the commit history of that branch largely intact, but I'm happy to squash if you'd prefer. A single commit named `Add nested belongs_to`?

deivid-rodriguez commented:
No, I'm prefer leaving the commit history, but the commit I'm pointing to was added by yourself and is no longer need.

rdallasgray commented:
@deivid-rodriguez OK, I've rebased correctly against upstream/master, which removes the redundant commit(s).

varyonic commented:
Hah!  I should have warned you, you just overwrote [my fixes last night](https://travis-ci.org/activeadmin/activeadmin/builds/194668366).

rdallasgray commented:
@varyonic Oh, dammit. Sorry. I wondered why you'd canceled the Travis build. How could a rebase against upstream have done that?

varyonic commented:
I think you force pushed to your GitHub repo.  David was just discussing force-with-lease with me.  Do you mind if I redo your rebase?

rdallasgray commented:
@varyonic Yeah, I had to force-push because of the rebase. I hadn't heard of force-with-lease -- that's useful. And no, I don't mind if you redo the rebase.

What I don't understand is how you made changes to a branch in my repo though?

deivid-rodriguez commented:
@rdallasgray Github allows you to choose whether the branch can be edited by the target project maintainers when you open it. I think the checkbox is pretty recent, so if you don't remember it maybe they applied it to old PR's when they enabled it? ¯\_(⊙︿⊙)_/¯

rdallasgray commented:
@deivid-rodriguez Interesting! That's definitely a bit counter-intuitive. Oh well, live and learn. Thanks for your patience.

varyonic commented:
FYI I did merge Bernado's commits, the history did not appear so useful.

rdallasgray commented:
Thanks @varyonic @Fivell @deivid-rodriguez @timoschilling and @bamorim!

varyonic commented:
@timoschilling  [#3280] [] [bamorim] should be added to CHANGELOG?

timoschilling commented:
@varyonic yes please

Fivell commented:
related to #4759

Fivell commented:
@rdallasgray ,  did you try using this solution to edit/create record what has multiple belongs_to ?

rdallasgray commented:
@Fivell My understanding of inherited_resources is that you can't do that -- multiple `belongs_to` declarations result in nested associations, I think?

Fivell commented:
@rdallasgray , what if one of them is optional ?

rosendi commented:
@Fivell could you please tell what is the current workaround for nesting belongs_to?

Fivell commented:
@rosendi  to be honest i don't know,  i always avoided to use nested belongs_to and using just a regular belongs_to instead
